File: Data Structures & Algorithms/valid-binary-search-tree/submission-2.kt
/**
 * Definition for a binary tree node.
 * class TreeNode(var `val`: Int) {
 *     var left: TreeNode? = null
 *     var right: TreeNode? = null
 * }
 */

class Solution {
fun isValidBST(root: TreeNode?): Boolean {
        return isValidSubTree(root, Long.MIN_VALUE, Long.MAX_VALUE)
    }

    private fun isValidSubTree(node: TreeNode?, min: Long, max: Long): Boolean {
        if (node == null) return true
        
        val currentVal = node.`val`.toLong()
        if (currentVal <= min || currentVal >= max) {
            return false
        }
        
        return isValidSubTree(node.left, min, currentVal) && 
               isValidSubTree(node.right, currentVal, max)
    }
}

File: Data Structures & Algorithms/kth-smallest-integer-in-bst/submission-2.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def kthSmallest(self, root: Optional[TreeNode], k: int) -> int:
        curr = k
        val = -1
        def dfs(node: Optional[TreeNode]) -> None:
            nonlocal curr
            nonlocal val

            if not node:
                return
            
            dfs(node.left)
            curr -= 1
            if curr == 0:
                val = node.val

            dfs(node.right)
        
        dfs(root)

        return val